src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageParser.cs

```diff
@@ -2,44 +2,70 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Binary;
 
 namespace Microsoft.AspNetCore.SignalR.Internal.Formatters
 {
     public static class BinaryMessageParser
     {
+        private static int[] _numBitsToShift = new[] { 0, 7, 14, 21, 28 };
+        private const int MaxLengthPrefixSize = 5;
+
         public static bool TryParseMessage(ref ReadOnlyBuffer<byte> buffer, out ReadOnlyBuffer<byte> payload)
         {
-            long length = 0;
-            payload = default(ReadOnlyBuffer<byte>);
+            payload = default;
 
-            if (buffer.Length < sizeof(long))
+            if (buffer.IsEmpty)
             {
                 return false;
             }
 
-            // Read the length
-            length = buffer.Span.Slice(0, sizeof(long)).ReadBigEndian<long>();
+            // The payload starts with a length prefix encoded as a VarInt. VarInts use the most significant bit
+            // as a marker whether the byte is the last byte of the VarInt or if it spans to the next byte. Bytes
+            // appear in the reverse order - i.e. the first byte contains the least significant bits of the value
+            // Examples:
+            // VarInt: 0x35 - %00111001 - the most significant bit is 0 so the value is %x0111001 i.e. 0x35 (53)
```

Comment:
Isn't 0x35 - `00110101`?

Reply:
Yes. (I am not fixing this today though - tomorrow will be a better day 😉 )